packages/core/admin/admin/src/pages/SettingsPage/pages/ApiTokens/EditView/index.js

```diff
@@ -294,13 +301,20 @@ const ApiTokenCreateView = () => {
           initialValues={{
             name: apiToken?.name || '',
             description: apiToken?.description || '',
-            type: state.selectedActions.length > 0 && !apiToken?.type ? 'custom' : apiToken?.type,
-            lifespan: apiToken?.lifespan,
+            type: apiToken?.type,
+            lifespan: apiToken?.lifespan ? apiToken.lifespan.toString() : apiToken?.lifespan,
           }}
           enableReinitialize
           onSubmit={(body, actions) => handleSubmit(body, actions)}
         >
           {({ errors, handleChange, isSubmitting, values }) => {
+            if (state.selectedActions.length > 0 && !values?.type) {
+              typeApiToken.current = typeApiToken.current ? typeApiToken.current : 'custom';
+              values.type = typeApiToken.current;
+            } else if (state.selectedActions.length === 0 && typeApiToken.current !== undefined) {
+              values.type = typeApiToken.current;
```

Comment:
Sorry to bring this up again, but it's the same pattern (mutations in render) as the previous commit, which I'm not sure is recommended. Maybe you could try extracting it in a subcomponent, so that you could do that logic outside the return, or in an effect for example

I'd love to get a second opinion on this topic though, as it may not be a problem, especially since JS found no issues in testing

Reply:
hey Remi I have refactored again the code now using a Formik features called setFieldValue https://formik.org/docs/api/formik#setfieldvalue-field-string-value-any-shouldvalidate-boolean--void and I use this utility only when the Permissions context change and the type is empty

with this solution I have fixed all the bugs.
Related to your suggestion, unfortunately the functions triggered when we change the Permissions are outside the Formik Component because are related to the ApiTokenPermissionsContextProvider, and this component wraps the Formik form and others components